pkg/reconcilermanager/controllers/helm_value_files.go

```diff
@@ -0,0 +1,165 @@
+		})
+	}
+	return cmsCMRefs
+}
+
+// getReconcilerHelmConfigMapRefs returns a list of ValuesFileRefs with the
+// names of the HelmValuesFile ConfigMap copies in the config-management-system
+// namespace with the associated data key.
+func (r *RepoSyncReconciler) getReconcilerHelmConfigMapRefs(rs *v1beta1.RepoSync) []v1beta1.ValuesFileRef {
+	if rs.Spec.Helm == nil {
+		return nil
+	}
+	var cmsCMRefs []v1beta1.ValuesFileRef
+	for _, vfRef := range rs.Spec.Helm.ValuesFileRefs {
+		copyCMRef := getHelmConfigMapCopyRef(vfRef.Name, client.ObjectKeyFromObject(rs))
+		cmsCMRefs = append(cmsCMRefs, v1beta1.ValuesFileRef{
+			Name:    copyCMRef.Name,
+			DataKey: validate.HelmValuesFileDataKeyOrDefault(vfRef.DataKey),
+		})
+	}
+	return cmsCMRefs
+}
+
+// upsertHelmConfigMaps creates or updates the helm values file ConfigMaps
+// in the config-management-system namespace using existing ConfigMaps in the
+// RepoSync namespace.
+func (r *RepoSyncReconciler) upsertHelmConfigMaps(ctx context.Context, rs *v1beta1.RepoSync, labelMap map[string]string) error {
+	if rs.Spec.SourceType != string(v1beta1.HelmSource) || rs.Spec.Helm == nil || len(rs.Spec.Helm.ValuesFileRefs) == 0 {
+		return nil
+	}
+	rsRef := client.ObjectKeyFromObject(rs)
+	for _, vfRef := range rs.Spec.Helm.ValuesFileRefs {
+		userCMRef := types.NamespacedName{
+			Namespace: rsRef.Namespace,
+			Name:      vfRef.Name,
+		}
+		copyCMRef := getHelmConfigMapCopyRef(userCMRef.Name, rsRef)
+		userCM, err := r.getUserHelmConfigMap(ctx, userCMRef)
+		if err != nil {
+			return errors.Wrapf(err, "user config map required for helm values: %s", userCMRef)
+		}
+		if _, err = r.upsertHelmConfigMap(ctx, copyCMRef, rsRef, userCM, labelMap); err != nil {
+			return err
+		}
+
+	}
+	// Since the ValuesFileRefs may have changed, we need to delete the copies
+	// of any that were removed.
+	return r.deleteUnusedHelmConfigMapCopies(ctx, rs)
+}
+
+// getUserHelmConfigMap gets a user managed ConfigMap in the same namespace as
+// the RepoSync.
+func (r *RepoSyncReconciler) getUserHelmConfigMap(ctx context.Context, userCMRef types.NamespacedName) (*corev1.ConfigMap, error) {
+	userCM := &corev1.ConfigMap{}
+	if err := r.client.Get(ctx, userCMRef, userCM); err != nil {
+		if apierrors.IsNotFound(err) {
+			return userCM, errors.Errorf(
+				"config map not found: %s", userCMRef)
+		}
+		return userCM, errors.Wrapf(err,
+			"config map get failed: %s", userCMRef)
+	}
+	return userCM, nil
+}
+
+// upsertHelmConfigMap creates or updates a config map in the
+// config-management-system namespace using an existing config map.
+func (r *RepoSyncReconciler) upsertHelmConfigMap(ctx context.Context, cmsCMRef, rsRef types.NamespacedName, userCM *corev1.ConfigMap, labelMap map[string]string) (controllerutil.OperationResult, error) {
+	cmsConfigMap := &corev1.ConfigMap{}
+	cmsConfigMap.Name = cmsCMRef.Name
+	cmsConfigMap.Namespace = cmsCMRef.Namespace
+
+	op, err := CreateOrUpdate(ctx, r.client, cmsConfigMap, func() error {
```

Comment:
Can we actually do an update? https://github.com/GoogleContainerTools/kpt-config-sync/blob/v1.16.0-rc.2/pkg/reconcilermanager/controllers/reposync_controller.go#L1249 mentions that the copied ConfigMap is immutable.

Reply:
I guess it depends if the metadata is immutable or just the data. The docs do indicate that watches are closed when immutable, so maybe the metadata is immutable too.

Then again, if something changed such that it's not semantically equal any more and it tries to update it, we would want to see that error, right?  CreateOrUpdate already skips updates if nothing changed.

Reply:
`cmsConfigMap.Immutable` only sets the data to be immutable. Its metadata can still be modified.
If the copiedConfigMap's metadata is accidentally updated, this will revert the change, which is the correct behavior.